modules/calib3d/src/solvepnp.cpp

```diff
@@ -456,4 +398,325 @@ int solveP3P( InputArray _opoints, InputArray _ipoints,
     return solutions;
 }
 
+int solvePnPGeneric( InputArray _opoints, InputArray _ipoints,
+                     InputArray _cameraMatrix, InputArray _distCoeffs,
+                     OutputArrayOfArrays _rvecs, OutputArrayOfArrays _tvecs,
+                     bool useExtrinsicGuess, int flags,
+                     InputArray _rvec, InputArray _tvec,
+                     OutputArray reprojectionError) {
+    CV_INSTRUMENT_REGION();
+
+    Mat opoints = _opoints.getMat(), ipoints = _ipoints.getMat();
+    int npoints = std::max(opoints.checkVector(3, CV_32F), opoints.checkVector(3, CV_64F));
+    CV_Assert( ( (npoints >= 4) || (npoints == 3 && flags == SOLVEPNP_ITERATIVE && useExtrinsicGuess) )
+               && npoints == std::max(ipoints.checkVector(2, CV_32F), ipoints.checkVector(2, CV_64F)) );
+
+    if( flags != SOLVEPNP_ITERATIVE )
+        useExtrinsicGuess = false;
+
+    if (useExtrinsicGuess)
+        CV_Assert( !_rvec.empty() && !_tvec.empty() );
+
+    int mtype = CV_64FC1;
+    if( useExtrinsicGuess )
+    {
+        int rtype = _rvec.type(), ttype = _tvec.type();
+        Size rsize = _rvec.size(), tsize = _tvec.size();
+        CV_Assert( (rtype == CV_32FC1 || rtype == CV_64FC1) &&
+                   (ttype == CV_32FC1 || ttype == CV_64FC1) );
+        CV_Assert( (rsize == Size(1, 3) || rsize == Size(3, 1)) &&
+                   (tsize == Size(1, 3) || tsize == Size(3, 1)) );
+    }
+    //TODO:
+//    else
+//    {
+//        // use CV_32F if all PnP inputs are CV_32F and outputs are empty
```

Comment:
> use CV_32F if all PnP inputs are CV_32F and outputs are empty

EPnP should return always in double precision:
https://github.com/opencv/opencv/blob/1f517b8a0293f2574c3dce739e5b273164244cac/modules/calib3d/src/epnp.cpp#L195-L196

Same thing with most of the methods, e.g. for P3P:
https://github.com/opencv/opencv/blob/1f517b8a0293f2574c3dce739e5b273164244cac/modules/calib3d/src/p3p.cpp#L55-L56

For iterative:
https://github.com/opencv/opencv/blob/1f517b8a0293f2574c3dce739e5b273164244cac/modules/calib3d/src/calibration.cpp#L1251-L1257

---

For me, even if all the input are in `CV_32F`, `rvec` and `tvec` should be overwritten in double precision.

I have changed the code to return `rvec` and `tvec` always in `CV_64F` depth. Internally, all these methods should use double for calculation.

Documentation should be updated to reflect that.

Reply:
> rvec and tvec should be overwritten in double precision

Not always.
Use this pattern:
```
if (R.needed()) _R.convertTo(R, R.empty() ? CV_64FC1 : R.type());
```
(however it is buggy in some cases - doesn't support empty `Mat_<float>`, but should raise an exception)

>  Internally, all these methods should use double for calculation.

This is the right way.

Reply:
This makes sense:
 - always return in `CV_64FC1`
 - except if `rvec` or `tvec` is not empty

Need to check that it works properly.